arches/app/models/graph.py

```diff
@@ -570,12 +570,15 @@ def delete_instances(self, verbose=False):
         """
         if verbose is True:
             bar = pyprind.ProgBar(Resource.objects.filter(graph_id=self.graphid).count())
+        delete_count = 0
         for resource in Resource.objects.filter(graph_id=self.graphid):
             resource.delete()
+            delete_count += 1
```

Comment:
While here, I'd prefer to get rid of the extra query we're doing to get the count.

IMO we should only define the queryset once, get its length, and then iterate over the already evaluated queryset.

Reply:
I had thought that django did some wizardry regarding optimized memory management in iterating through a queryset but after looking into it, that doesn't seem to be the case (i.e. nothing distinct from lazy query evaluation). Otherwise I would have just declared the queryset as a list from the get-go (which is what I've done in the latest commits)

Reply:
> I had thought that django did some wizardry regarding optimized memory management in iterating through a queryset but after looking into it

That's right, it will reuse the objects if you iterate an evaluated queryset, but it only works if you reuse the queryset, meaning you store a reference to it in a variable. The issue (predating your changes here of course) was that when you build it again from scratch with .objects instead, it's a new query.

If I'm missing your point, though, say so! :D

Reply:
Oh, if your point is: why not just always convert to list?

I guess if you want to do something else like batch delete or further filtering, you can leave it as a queryset and queryset.delete(), which we can't do here without removing the progress bar.

So no harm no foul to listify it here, but there is something nice about staying a little future-proof and leaving as a queryset if you can. QoL for the next person touching the code, potentially.